repos/system_upgrade/common/libraries/grub.py

```diff
@@ -33,39 +230,126 @@ def blk_dev_from_partition(partition):
     try:
         result = run(['lsblk', '-spnlo', 'name', partition])
     except CalledProcessError:
-        api.current_logger().warning(
-            'Could not get parent device of {} partition'.format(partition)
-        )
-        raise StopActorExecution()  # TODO: return some meaningful value/error
+        msg = 'Could not get parent device of {} partition'.format(partition)
+        api.current_logger().warning(msg)
+        raise StopActorExecution(msg)
+
     # lsblk "-s" option prints dependencies in inverse order, so the parent device will always
     # be the last or the only device.
     # Command result example:
     # 'result', {'signal': 0, 'pid': 3872, 'exit_code': 0, 'stderr': u'', 'stdout': u'/dev/vda1\n/dev/vda\n'}
     return result['stdout'].strip().split()[-1]
 
 
+def _get_partition(directory):
+    """
+    Get partition name of `directory`.
+    """
+
+    try:
+        result = run(['grub2-probe', '--target=device', directory])
+    except CalledProcessError:
+        msg = 'Could not get name of underlying {} partition'.format(directory)
+        api.current_logger().warning(msg)
+        raise StopActorExecution(msg)
+    except OSError:
+        msg = ('Could not get name of underlying {} partition:'
+               ' grub2-probe is missing.'
+               ' Possibly called on system that does not use GRUB2?').format(directory)
+        api.current_logger().warning(msg)
+        raise StopActorExecution(msg)
+
+    partition = result['stdout'].strip()
+    api.current_logger().info('{} is on {}'.format(directory, partition))
+
+    return partition
+
+
 def get_boot_partition():
     """
     Get /boot partition name.
     """
+
+    return _get_partition('/boot')
+
+
+def is_efi():
+    """
+    Return True if UEFI is used.
+
+    NOTE(pstodulk): the check doesn't have to be valid for hybrid boot (e.g. AWS, Azure, ..)
+    """
+
+    return os.path.exists("/sys/firmware/efi")
+
+
+def get_efi_partition():
+    """
+    Return the EFI System Partition (ESP).
+
+    Raise StopActorExecution when:
+        - UEFI is not detected,
+        - ESP is not mounted where expected,
+        - the partition can't be obtained from GRUB.
+    """
+
+    if not is_efi():
+        raise StopActorExecution('Unable to get ESP when BIOS is used.')
+
+    if not os.path.exists(EFI_MOUNTPOINT) or not os.path.ismount(EFI_MOUNTPOINT):
+        raise StopActorExecution(
+            'The UEFI has been detected but the ESP is not mounted in /boot/efi as required.'
+        )
+
+    return _get_partition('/boot/efi/')
+
+
+def get_blk_device(device):
```

Comment:
How is this different from the existing `blk_dev_from_partition()` function?
It could use the docstring though :)

Reply:
another refactoring leftover. Thanks!